### ISSUE TITLE
ETIMEDOUT in many req in loop

### ISSUE THREAD

**Issue by saeedhei** [axios/axios#2997]
Hello

Why I'm getting ETIMEDOUT in many request in loop?

**Example Code**
Code snippet to illustrate your question

```js
// Example code here
const axios = require('axios');
async function getUser() {
  try {
    const response = await axios.get('http://example.com/projects/book/api/view/');
    console.log(response);
  } catch (error) {
    console.error(error);
  }
}

var j;
for (j = 0; j < 350; j++) {
  getUser()
}
```

Error:
```
{ Error: connect ETIMEDOUT x.x.x.x:80
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1113:14)
  errno: 'ETIMEDOUT',
  code: 'ETIMEDOUT',
  syscall: 'connect',
  address: 'x.x.x.x',
  port: 80,
  config:
   { url: 'http://example.com/projects/book/api/view/',
     method: 'get',
     headers:
      { Accept: 'application/json, text/plain, */*',
        'User-Agent': 'axios/0.19.0' },
     transformRequest: [ [Function: transformRequest] ],
     transformResponse: [ [Function: transformResponse] ],
     timeout: 0,
     adapter: [Function: httpAdapter],
     xsrfCookieName: 'XSRF-TOKEN',
     xsrfHeaderName: 'X-XSRF-TOKEN',
     maxContentLength: -1,
     validateStatus: [Function: validateStatus],
     data: undefined },
```

**Comment by michaelherger** [comment]
I can easily reproduce that issue. Every now and then one of the requests would fail. Sometimes I even got the random `ENOTFOUND`. "Serializing" the requests by awaiting the first response before sending the next one would not show that behaviour. But that's not a real world solution.

**Comment by abettke** [comment]
Also seeing similar behavior when running multiple requests using `Promise.all()`. The `ETIMEDOUT` also seems to be sporadic, it doesn't occur all the time.

**Comment by jasonsaayman** [comment]
This looks to not be an issue directly with Axios. If I am incorrect please open a new issue and provider more information.

**Comment by therockstorm** [comment]
If this isn't a problem with `axios`, what is it a problem with? I'm seeing the same issue with version `0.21.4`.

**Comment by andyias** [comment]
@jasonsaayman same issue with  version 0.25.0
still not an issue directly with axios ?

**Comment by jasonsaayman** [comment]
@andyias yeah for sure not an issue with axios if you are referring to the initial code snippet, I mean just of the top of my head the things that could cause a timeout:

- Rate limiting
- The endpoints ability to handle x amount of request per x amount of execution time
- The loop could be causing your app itself to crash or timeout etc.

**Comment by andyias** [comment]
@jasonsaayman  right
for anyone experiencing this same issue, try splitting your tasks into smaller  chunks and execute them in parallel

**Comment by webb24h** [comment]
This shouldn't be closed. Still experiencing same bug when using axios in a forEach loop

**Comment by zeuspsy0x1** [comment]
For anyone out there that is completely sure that the code is ok. Try switching internet networks. I don't know why, but I got this error and the only solution was switching to another internet.

**Comment by anogray** [comment]
> @andyias yeah for sure not an issue with axios if you are referring to the initial code snippet, I mean just of the top of my head the things that could cause a timeout:
> 
> - Rate limiting
> - The endpoints ability to handle x amount of request per x amount of execution time
> - The loop could be causing your app itself to crash or timeout etc.

I have to update data in db say 5k docs quickly and I am sending 500 requests parallelly kind of nodejs Promised , with axios ans 500 requests needs cooldown of some time and then start next iteration.

Sometimes I get Etimeout sometimes I do not. Same goes in localhost, azure functions, lambdas, all.
Though keep alive with request and timeout axios duration configurations reduced the number of Etimeouts but still the issue persists.
Is it something with axios or what could be , rate limit to the endpoint won't be as it is specific paid endpoint they told.
Also the loop/code actually doesn't crash.
I am more than 2 weeks no idea to figure out reason.

**Comment by anogray** [comment]
> This shouldn't be closed. Still experiencing same bug when using axios in a forEach loop

It is not only with for each loop any other iteration like for too..

**Comment by JoeKaldas** [comment]
> For anyone out there that is completely sure that the code is ok. Try switching internet networks. I don't know why, but I got this error and the only solution was switching to another internet.

I can say this is happening with me. It doesn't work on my work/home network but works perfectly fine when I hotspot.
It works though on my friend's same network at work but he's on Windows and I'm on Mac.

**Comment by wilnami** [comment]
this also happened to me

**Comment by zeuspsy0x1** [comment]
In the case of this issue the promise should keep waiting for data without throwing an error (unless you specify a limit of time for the promises). In the end, switching internet networks works for this error because I'm 99% sure it's not a code error. It's an Internet service Provider error. You just have to call your ISP and tell them to check why the http requests are failing. Tell them that and they will know what to do (I don't have a clue why, but they are the only ones with the power to solve it).

**Comment by iluxonchik** [comment]
I'm looking into a similar issue, and the core of the problem seems to be that you are running out of outbound ports. The solution is to set `keepAlive: true` and add an explicit `timeout` on axios.

For more information, refer to the answer on [this StackOverflow question](https://stackoverflow.com/questions/63064393/getting-axios-error-connect-etimedout-when-making-high-volume-of-calls)

**Comment by botzill** [comment]
Hi @iluxonchik. Can you elaborate more on outbound ports issue? Is this smith that can be fixed increased ?

**Comment by phapntm** [comment]
> I'm looking into a similar issue, and the core of the problem seems to be that you are running out of outbound ports. The solution is to set `keepAlive: true` and add an explicit `timeout` on axios.
> 
> For more information, refer to the answer on [this StackOverflow question](https://stackoverflow.com/questions/63064393/getting-axios-error-connect-etimedout-when-making-high-volume-of-calls)

might be you are right. 
But axios can not use `keepAlive: true` with proxy. So the problems still persists.
